docs/design-documents/20230228-connector-lifecycle-methods.md

```diff
@@ -0,0 +1,226 @@
+is invalid) but we still executed the initialization. On the other hand,
+if `Configure` was already called the lifecycle methods don't have to parse the
+config itself and they can know that the config is valid. That's why we should
+call lifecycle methods _after_ `Configure` and before `Open` (at that point the
+initialization already needs to be done).
+
+The lifecycle methods should be called at the following points:
+
+- Connector creation - should be called the first time the pipeline is started
+  after the connector was created.
+- Connector update - should be called the first time the pipeline is started
+  after the connector was updated (i.e. it was started once before already).
+- Connector deletion - should be called when the connector gets deleted. If it
+  fails, the error _should not_ prevent the deletion of the connector.
+
+We need to define which configuration gets passed to lifecyle methods. For this
+let's define that a connector configuration is "active" only after a lifecycle
+method was called with that configuration and no error was returned.
+Consequently, there are some edge cases we need to look out for:
+
+- If a new connector is created and deleted without being started in between,
+  the configuration is not treated as "active" so the deletion lifecycle method
+  won't be called because the creation was not triggered either (it would only
+  be triggered if the connector was actually started).
+- If a new connector is created and updated without being started in between,
+  the configuration passed to the creation lifecycle method should be the latest
+  one (after the update).
+- If a connector is updated twice without being started between updates, the
+  update lifecycle method should be called with the last active configuration
+  (old) and with the latest configuration (new).
+- If a connector is updated and deleted without being started in between, the
+  deletion lifecycle method should receive the old configuration before the
+  update.
+- If a connector is not configured correctly (i.e. `Configure` returns an
+  error), then no lifecycle method is called and that configuration should not
+  be treated as active (unless it was already active before).
+
+If a pipeline is provisioned using a pipeline configuration file we need to make
+sure to call the correct connector lifecycle method based on if the
+configuration actually changed (we need to make a diff). Similarly, if a
+connector gets removed from the configuration file or the configuration file is
+removed entirely, the deletion lifecycle method should be called (given that the
+connector was active before).
+
+We need to make sure that Conduit stays backwards compatible with existing
+standalone connectors, which means we need to catch errors in case the newly
+introduced methods in the connector protocol do not exist.
+
+### Recommendation
+
+Option 1 seems like the most straightforward option with enough freedom to
+evolve the design in the future.
```

Comment:
Agreed :+1:

Reply:
+1

Reply:
👍 agree as well. while the other options are less verbose, they lack the clarity.

Reply:
agreed - option 1 looks most clear to develop against!

Reply:
one more for option 1, i think being explicit is going to provide value in the feature rather than jamming it all in the same method.